### PR TITLE
Documentation updates + minor code fixes

### PR DESCRIPTION
**Pull request by fidian**
* Allowing AWS credentials in config.
* AWS credentials are optional.
* Fixing example scripts to work with the renamed container.
* Updating the test script to not run coverage if not under source
control.
* Fixing up some places where I noticed the style guide was not followed.
* It's not "API keys", it is "Access Codes" or "Access Code Pair".
* Oh, and documentation.

### CODE REVIEW

**Comment by fidian** [comment]
DOT-4
Merge, don't squash.

File: doc/request-signatures.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+Request Signatures
+==================
+
+In order to verify that a request is being made with a valid Access Code Pair and to prevent forgeries, a request signature is used.  The Access Code Pair refers to both the Public Code and the Secret Code.  The Public Code is sent back to the API for identification, similar to a username.  The Secret Code is never sent back to the API and is strictly used for signature generation.
```
Comment (AbsentSemicolon) [inline doc/request-signatures.md]: Can drop that after verify.

File: doc/storage.md [hunk 2]
```diff
@@ -1,21 +1,41 @@
...
-            "region": "us-east-1"
+            "region": "us-east-1",
+            "secretAccessKey": "your secret access key, optional"
         }
     }
 
-### S3
+### `storage.engine`
+
+Determines the storage backend to use.  Only `s3` is supported.
```
Comment (AbsentSemicolon) [inline doc/storage.md]: We should say "Currently only `s3` is supported."

File: doc/config-common.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+PBKDF2 defines a way for a hashing algorithm to consume more time by running the hash repeatedly.  It is supposed to slow down hackers and is configurable so you can tweak the algorithm to use more time as computers get faster.
+
+This type of hashing lets you adjust the hashing `algorithm` that is used and the `iterations` for how many times to encode.  When done, it generates `derivedLength` bytes of binary data and that is later encoded using the specified `encoding`.
+
+
+`hmac`
+------
+
+This is a cryptographic hash signature.  The `algorithm` and `digest` must both be listed in [lib/ciphers-and-hashes.js].  The hash algorithm is seeded with data that is hashed with the `digest` a number of `iterations`.  This generates an initial key and the `algorithm` is used to generate a signature.
```
Comment (fidian) [inline doc/config-common.md]: Ignoring this one.

**Comment by fidian** [comment]
Added tests.  Made changes to cover all of your comments.

Do not merge.  If this is approved I will squash and then someone else can merge.

File: doc/record-encryption.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+
+Encryption Process
+------------------
+
+Incoming data is known as the "plaintext".  This does not mean the client is forced to send us unencrypted information.  Plaintext means it is the desired information when we are done decrypting.
+
+The plaintext is encrypted according to the [encryption config].  This is the ciphertext.  At the same time, the encryption options are built for this record.  Those options are saved in the record so it can be decrypted later, even if the [encryption config] changes.  The encryption requires an initialization vector (IV), which is simply random data.  It is used to pick a random starting position for the cipher.
```
Comment (fidian) [inline doc/record-encryption.md]: Yes.

File: doc/record-encryption.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+
+Layered Encryption
+------------------
+
+The technique of layered encryption is where you take some information and encrypt it multiple times.  You can provide the encryption keys to separate parties in order to keep the encrypted data more secure.  A different approach is to use a single secret that you build from enough information from various people; see [Shamir's Secret Sharing](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing).
+
+Layered encryption provides benefits.  For instance, PCI DSS 3.2 has some rules governing key rotation in section 3.6 that could apply.  The service is able to decrypt the outer layer and rekey the stored information.  Also, some information may be acceptable to be seen again without decrypting the entire record.  As an example, the account record that is stored holds the MFA key and hashed password in the inner layer to be safe.  The outer layer encrypts the email address, so emails to the user may be sent before their password expires.
```
Comment (fidian) [inline doc/record-encryption.md]: Can't link.  It's behind a non-distribution agreement.

Disagreed with the comma.  Preferred rewording of the sentence.

File: doc/email.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+
+### `email.ses.accessKeyId` and `email.ses.secretAccessKey`
+
+Your SES keys for AWS, used for communicating directly with the SES endpoint inside AWS.
+
+
+### `email.ses.region`
+
+What SES region you wish to use.
```
Comment (fidian) [inline doc/email.md]: I would prefer to not list the regions.  They keep growing.

File: doc/config.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+
+
+### `account.passwordHash`
+
+This configuration is sent to the client when registering for the service.  The user will hash their password according to this spec and send it back to us.  See `hash` in [Common Data Structures].
+
+
+### `account.passwordSaltLength`
+
+A random salt that is used to seed the secure hash generation function.  The generation function makes an ID of the necessary length and calls it the password salt.
```
Comment (fidian) [inline doc/config.md]: Clarified - "password's salt".

File: doc/config.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+
+### `accessCode.lifetime`
+
+How long the access code should live until it is expired.  See `lifetime` in [Common Data Structures].  PCI-compliant services will have this set to no more than 6 months.
+
+
+### `accessCode.secretLength`
+
+When generating random access codes, the private portion will be this many bytes long.
```
Comment (fidian) [inline doc/config.md]: Ends in preposition.  Used alternate wording.

**Comment by fidian** [comment]
@AbsentSemicolon @lnagle93 Any chance I could get you two to sign off on this ticket?  I know one of you already did in our chat room but I would prefer it on the ticket.

Or should I squash first and then you both can check?  Maybe you should confirm the commit I'm on and watch me squash.  It would be easier than auditing the code again.

**Comment by lnagle93** [comment]
Changes look good to me.

**Comment by AbsentSemicolon** [comment]
Reviewed and looks good.

**Pull request merged**